### PR TITLE
fix: Surface all auth server errors during login in email input.

### PR DESCRIPTION
**Pull request by SokratisVidros**
### What change does this PR introduce?

This PR patches the following issue:

1. Sign up with Github with user A
2. Sign out
3. Try to sign in with the email of user A
4. The API returns a 400 error not displayed in the Login form right away in smaller screens.

### Why was this change needed?

After this PR, all server auth errors will be displayed in the UI.

It also contains some minor copywriting fixes and clean-ups.

#### Before

https://github.com/novuhq/novu/assets/1352422/674ac4ef-7914-419d-bf48-99d3e163a752

#### After

Screenshot 2024-03-20 at 00 39 54

### CODE REVIEW

File: apps/api/src/app/auth/e2e/login.e2e.ts [hunk 1]
```diff
@@ -13,154 +13,179 @@ describe('User login - /auth/login (POST)', async () => {
...
       expect(body.statusCode).to.equal(401);
-    }
+      expect(body.message).to.contain('Account blocked');
+    });
+  });
 
-    const { body } = await session.testAgent.post('/v1/auth/login').send({
-      email: userCredentials.email,
-      password: userCredentials.password,
+  context('with OAuth', async () => {
```
Comment (SokratisVidros) [inline apps/api/src/app/auth/e2e/login.e2e.ts:169]: All the above ☝️ are just indentation changes after adding a `context`.

File: apps/web/src/pages/auth/components/LoginForm.tsx [hunk 2]
```diff
@@ -78,29 +79,28 @@ export function LoginForm({ email, invitationToken }: LoginFormProps) {
     }
   };
 
-  const serverErrorString = useMemo<string>(() => {
-    return Array.isArray(error?.message) ? error?.message[0] : error?.message;
-  }, [error]);
+  const emailClientError = errors.email?.message;
+  let emailServerError = parseServerErrorMessage(error);
```
Comment (SokratisVidros) [inline apps/web/src/pages/auth/components/LoginForm.tsx:83]: Similar changes can be applied to the register flow.

File: apps/api/src/app/auth/usecases/login/login.usecase.ts [hunk 3]
```diff
@@ -41,7 +41,8 @@ export class Login {
       throw new UnauthorizedException(`Account blocked, Please try again after ${blockedMinutesLeft} minutes`);
     }
 
-    if (!user.password) throw new ApiException('OAuth user login attempt');
+    // TODO: Trigger a password reset flow automatically for existing OAuth users instead of throwing an error
+    if (!user.password) throw new ApiException('Please sign in using Github.');
```
Comment (SokratisVidros) [inline apps/api/src/app/auth/usecases/login/login.usecase.ts:45]: Note that returning different errors per auth flow results in User enumeration issues. User enumeration is mostly a privacy concern and should be tackled in a separate PR.

Comment (SokratisVidros) [reply apps/api/src/app/auth/usecases/login/login.usecase.ts:45]: Moreover, the error message mentions Github specifically. Currently, it's the only supported OAuth method. When more methods are added, the copywriting should be tweaked.

File: apps/api/src/app/auth/e2e/login.e2e.ts (hunk 1 above)
Comment (scopsy) [reply apps/api/src/app/auth/e2e/login.e2e.ts:169]: A new one for me! What is the difference between context and describe?

Comment (SokratisVidros) [reply apps/api/src/app/auth/e2e/login.e2e.ts:169]: `context` is a synonym for `describe` that is usually available in BDD style APIs. For example https://www.coreycleary.me/better-test-structuring-using-mochas-context

File: apps/api/src/app/auth/usecases/login/login.usecase.ts (hunk 3 above)
Comment (LetItRock) [reply apps/api/src/app/auth/usecases/login/login.usecase.ts:45]: Btw we have the Google OAuth logic but currently, it's enterprise-only and "disabled".

**Review by LetItRock — APPROVED** [review]
🙌
